File: StardewSeedSearcher/Core/HashHelper.cs
using System;
using System.Data.HashFunction.xxHash;
using System.Text;

namespace StardewSeedSearcher.Core
{
    /// <summary>
    /// 哈希和随机种子计算辅助类
    /// 基于 Underscore76 的实现
    /// </summary>
    public static class HashHelper
    {
        /// <summary>XXHash 32位哈希函数</summary>
        private static readonly IxxHash Hasher = xxHashFactory.Instance.Create(new xxHashConfig { HashSizeInBits = 32 });

        /// <summary>
        /// 获取字符串的确定性哈希值
        /// </summary>
        public static int GetHashFromString(string value)
        {
            byte[] data = Encoding.UTF8.GetBytes(value);
            return GetHashFromBytes(data);
        }

        /// <summary>
        /// 获取整数数组的确定性哈希值
        /// </summary>
        public static int GetHashFromArray(params int[] values)
        {
            byte[] data = new byte[values.Length * 4];
            Buffer.BlockCopy(values, 0, data, 0, data.Length);
            return GetHashFromBytes(data);
        }

        /// <summary>
        /// 获取字节数组的确定性哈希值
        /// </summary>
        private static int GetHashFromBytes(byte[] data)
        {
            byte[] hash = Hasher.ComputeHash(data).Hash;
            return BitConverter.ToInt32(hash, 0);
        }

        /// <summary>
        /// 计算随机种子
        /// 模拟 StardewValley.Utility.CreateRandomSeed()
        /// </summary>
        /// <param name="useLegacyRandom">是否使用旧随机模式</param>
        public static int GetRandomSeed(int a, int b, int c, int d, int e, bool useLegacyRandom)
        {
            // 确保参数在有效范围内
            a %= 2147483647;
            b %= 2147483647;
            c %= 2147483647;
            d %= 2147483647;
            e %= 2147483647;

            if (useLegacyRandom)
            {
                // 旧随机：简单相加取模
                long sum = (long)a + b + c + d + e;
                return (int)(sum % 2147483647);
            }
            else
            {
                // 新随机：使用 XXHash
                return GetHashFromArray(a, b, c, d, e);
            }
        }
    }
}

File: pkg/core/hash_helper.go
package core

import (
	"encoding/binary"
	"hash"

	"github.com/pierrec/xxHash/xxHash32"
)

// HashHelper 提供哈希和随机种子计算功能
// 基于 Underscore76 的实现
type HashHelper struct {
	hasher hash.Hash32
}

// NewHashHelper 创建新的 HashHelper 实例
func NewHashHelper() *HashHelper {
	return &HashHelper{
		hasher: xxHash32.New(0),
	}
}

// GetHashFromString 从字符串获取确定性哈希值
func (h *HashHelper) GetHashFromString(value string) int32 {
	data := []byte(value)
	return h.GetHashFromBytes(data)
}

// GetHashFromArray 从整数数组获取确定性哈希值
func (h *HashHelper) GetHashFromArray(values ...int) int32 {
	data := make([]byte, len(values)*4)
	for i, v := range values {
		binary.LittleEndian.PutUint32(data[i*4:], uint32(v))
	}
	return h.GetHashFromBytes(data)
}

// GetHashFromBytes 从字节数组获取确定性哈希值
func (h *HashHelper) GetHashFromBytes(data []byte) int32 {
	// 重置哈希器到初始状态
	h.hasher.Reset()
	h.hasher.Write(data)
	hash32 := h.hasher.Sum32()

	// 模拟C#的BitConverter.ToInt32行为
	// BitConverter.ToInt32将字节数组转换为有符号32位整数
	return int32(hash32)
}

// GetRandomSeed 计算随机种子
// 模拟 StardewValley.Utility.CreateRandomSeed()
func (h *HashHelper) GetRandomSeed(a, b, c, d, e int, useLegacyRandom bool) int {
	// 确保参数在有效范围内
	a = a % 2147483647
	b = b % 2147483647
	c = c % 2147483647
	d = d % 2147483647
	e = e % 2147483647

	if useLegacyRandom {
		// 旧随机：简单相加取模
		sum := int64(a) + int64(b) + int64(c) + int64(d) + int64(e)
		return int(sum % 2147483647)
	} else {
		// 新随机：使用 XXHash
		return int(h.GetHashFromArray(a, b, c, d, e))
	}
}

// 全局实例，方便使用
var DefaultHashHelper = NewHashHelper()

// 使用全局实例的便利函数（线程安全版本）
func GetHashFromString(value string) int32 {
	helper := NewHashHelper()
	return helper.GetHashFromString(value)
}

func GetHashFromArray(values ...int) int32 {
	helper := NewHashHelper()
	return helper.GetHashFromArray(values...)
}

func GetRandomSeed(a, b, c, d, e int, useLegacyRandom bool) int {
	helper := NewHashHelper()
	return helper.GetRandomSeed(a, b, c, d, e, useLegacyRandom)
}
